.grit/patterns/hathora_to_new_ts_sdk.md

```diff
@@ -0,0 +1,226 @@
+        and {
+          contains `const $binding = new $resource()` => `const $binding = sdk.$member_name;`, 
+          rewrite_method_calls(resource=$resource, callee=$binding),
+        },
+        contains type_annotation(type=$resource) where {
+          $source = resource_type_import_source($resource),
+          $substitute = resource_type_substitute($resource),
+          $substitute <: ensure_import_from(source=$source)
+      } => `: $substitute`
+    }
+  }
+}
+
+// v1 functions return an extra layer of indirection now, so eg a login response is at `.loginResponse`
+pattern replace_inside_functions($client_type) {
+  function_declaration(parameters=$params, body=$body) where {
+    // TODO critically, this depends on types. if the target project is pure js, we're SoL.
+    // One option is to just assume any name calling a given method like `getConnectionInfo` is using Hathora.
+    // Not every method name may be that distinctive, and in any real codebase it could be insanely noisy.
+    // A more invasive solution that seems like it would require Grit core support
+    // is just following variable refs by substitution (where possible!) and guessing what's what.
+    // There is a point at which that's just reproducing a static analyzer.
+    $params <: [..., `$client_binding: $client_type`, ...],
+    $body <: rewrite_method_calls(resource=$client_type, callee=$client_binding),
+  }
+}
+
+pattern gather_calls() {
+  and { 
+    any {
+      bubble any {
+        generic_replace_resource(resource=`RoomV1Api`),
+        replace_inside_functions(client_type=`RoomV1Api`)
+      },
+      bubble and {
+        $resource = `LobbyV1Api`,
+        any {
+          generic_replace_resource(resource=$resource),
+          replace_inside_functions(client_type=$resource)
+        }
+      },
+      bubble and {
+        $resource = `AuthV1Api`,
+        any {
+          generic_replace_resource(resource=$resource),
+          replace_inside_functions(client_type=$resource)
+        }
+      },
+      bubble and {
+        $resource = `LobbyV2Api`,
+        any {
+          generic_replace_resource(resource=$resource),
+          replace_inside_functions(client_type=$resource)
+        }
+      }
+    }
+  }
+}
+
+pattern type_imports() {
+  or {
+    `Region`,
+    `Lobby`
+  }
+}
+
+pattern replace_type_import($imported_type) {
+  import_statement(import = import_clause(name=named_imports($imports)), source= $old) where {
+    $old <: or {
+    `"@hathora/hathora-cloud-sdk"`,
+    `"@hathora/hathora-cloud-sdk/src/models/index"`
+  },
+    if ($imports <: contains $import where $import <: $imported_type) {
+      $imports <: contains $import,
+      $import <: ensure_import_from(source=`"@hathora/cloud-sdk-typescript/dist/sdk/models/shared"`)
+    } else {
+      true
+    }
+  },
+}
+
+any {
+  replace_type_import(imported_type=`Lobby`),
+  replace_type_import(imported_type=`Region`),
```

Comment:
This seems like it is going to be very verbose. Is there a problem with just using `replace_import` with a loose binding wherre we remap (everything imported from `"@hathora/hathora-cloud-sdk"` to `@hathora/cloud-sdk-typescript`.

Reply:
There's no problem 👍 (AFAIK; will go do it and find out)